recipes_source/recipes/benchmark.py

```diff
@@ -0,0 +1,828 @@
+#       Generic implementation.  |  7.9  |  8.7  |  8.9  |  9.3   |   11   |    15   |    24 
+#
+# Times are in microseconds (us).
+
+######################################################################
+# 6. Generating bechmark inputs with Fuzzed Parameters
+#
+# We'll take a brief detour and use fuzzed inputs to discuss transpose
+# and contiguous before returning to shift.
+#
+
+from torch.utils.benchmark import Fuzzer, FuzzedParameter, FuzzedTensor, ParameterAlias
+
+example_fuzzer = Fuzzer(
+    parameters = [
+        FuzzedParameter("k0", minval=1, maxval=1024 ** 2, distribution="loguniform"),
+        FuzzedParameter("k1", distribution={1: 0.5, ParameterAlias("k0"): 0.5}, strict=True),
+    ],
+    tensors = [
+        FuzzedTensor("x", size=("k0", "k1"), min_elements=128, max_elements=128 * 1024, probability_contiguous=0.6)
+    ],
+    seed=0,
+)
+
+results = []
+for tensors, tensor_params, params in example_fuzzer.take(10):
+    sub_label=f"{params['k0']:<6} x {params['k1']:<4} {'' if tensor_params['x']['is_contiguous'] else '(discontiguous)'}"
+    for stmt in ("x.contiguous()", "x.t().contiguous()"):
+        timer = Timer(
+            stmt,
+            globals=tensors, 
+            label="2D transpose",
+            description=stmt,
+            sub_label=sub_label,
+        )
+        results.append(timer.blocked_autorange())
+
+compare = Compare(results)
+compare.trim_significant_figures()
+compare.print()
+
+# [------------------------------- 2D transpose ------------------------------]
+#                                      |  x.contiguous()  |  x.t().contiguous()
+# 1 threads: ------------------------------------------------------------------
+#       355    x 355  (discontiguous)  |      200000      |          2000      
+#       751    x 1                     |         200      |          2000      
+#       313    x 313                   |         200      |        170000      
+#       45851  x 1                     |         200      |          2000      
+#       146    x 146                   |         200      |         43000      
+#       15854  x 1                     |         200      |          2000      
+#       143    x 143  (discontiguous)  |       40000      |          2000      
+#       2709   x 1                     |         200      |          2000      
+#       312    x 312                   |         200      |        170000      
+#       5674   x 1                     |         200      |          2000      
+#
+# Times are in nanoseconds (ns).
+
+######################################################################
+# The fuzzed benchmarks reveal several noteworthy features:
+#
+# If a Tensor is already contiguous we do not need to construct a new
+# Tensor and the operation is extremely cheap. O(100 ns)
+#
+# For N x 1 Tensors transpose requires that we create a new Tensor,
+# but we can reuse the same buffer.
```

Comment:
`buffer` is an unclear and overloaded word, "we don't have to do any physical data movement"